### PR TITLE
Add support for "User Default" settings II

### LINKED ISSUE #2325: Default Profile for Common Profile Settings
# Default Profile for Common Profile Settings

Allow settings which the user would like applied to all profiles to be specified in a profile with a well-known name, such as `"name" : "_default"`. All settings in this profile will be applied before they are overridden from individual profiles. This would be helpful to allow all profiles to have acrylic applied with a certain opacity where those settings are not specified.

    [snip]
    {
        "name" : "_default",
        "useAcrylic" : true,
        "acrylicOpacity" : 0.5,
        "closeOnExit" : true,
        "colorScheme" : "Campbell",
        "cursorColor" : "#FFFFFF",
        "cursorShape" : "bar",
        "fontFace" : "Consolas",
        "fontSize" : 10,
        "historySize" : 9001,
        "padding" : "0, 0, 0, 0",
        "snapOnInput" : true,
        "startingDirectory" : "%USERPROFILE%"
    },
    {
        "name" : "PowerShell Core",
        "commandline" : "C:\\Program Files\\PowerShell\\6\\pwsh.exe",
        "guid" : "{574e775e-4f2a-5b96-ac1e-a2962a402336}",
        "icon" : "ms-appx:///ProfileIcons/{574e775e-4f2a-5b96-ac1e-a2962a402336}.png",
    },
    {
        "name" : "cmd",
        "commandline" : "cmd.exe",
        "guid" : "{0caa0dad-35be-5f56-a8ff-afceeeaa6101}",
        "icon" : "ms-appx:///ProfileIcons/{0caa0dad-35be-5f56-a8ff-afceeeaa6101}.png",
        "useAcrylic" : false
    },
    [snip]

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

_This is attempt 2 at this feature_. The original PR can be found at #3369. 

These are settings that apply to _every_ profile, before user customizations. 

If the user wants to add "default profile settings", they can make the `"profiles"` property an _object_, instead of a list, and add `"defaults"` key underneath that object. The users list of profiles should then be under the `list` property of the `profiles` object.

## References
#2515, #2603, #3369, #3569

## PR Checklist
* [x] Closes #2325
* [x] I work here
* [x] Tests added/passed
* [x] schema, docs updated

## Detailed Description of the Pull Request / Additional comments
~~Discussion in #2325 itself serves as the "spec" for this task. I thought we'd need more discussion on the topic, but it ended up being pretty straightforward.~~

I should not have said that in the original PR. We've had a better spec review now that I think we're happier with.

## Validation Steps Performed
_ran the tests_

### CODE REVIEW

File: doc/user-docs/UsingJsonSettings.md [hunk 1]
```diff
@@ -199,6 +199,119 @@ like to hide all the WSL profiles, you could add the following setting:
 
 ```
 
+### Default settings
+
+In [#2325](https://github.com/microsoft/terminal/issues/2325), we introduced the
+concept of "Default Profile Settings". These are settings that will apply to all
+of your profiles by default. Profiles can still override these settings
+individually. With default profile settings, you can easily make changes to all
+your profiles at once. For example, lets say I have the following settings:
```
Comment (DHowett-MSFT) [inline doc/user-docs/UsingJsonSettings.md]:
```suggestion
your profiles at once. For example, given the following settings:
```

File: doc/user-docs/UsingJsonSettings.md [hunk 2]
```diff
@@ -199,6 +199,119 @@ like to hide all the WSL profiles, you could add the following setting:
...
+                "startingDirectory" : "%USERPROFILE%"
+            }
+        ],
+    }
+```
+
+Note that the `profiles` property has changed in this example from a _list_ of
+profiles, to an _object_ with two properties:
+* a `list` that contains the list of all the profiles
+* the new `defaults` object, that contains all the settings that should apply to
```
Comment (DHowett-MSFT) [inline doc/user-docs/UsingJsonSettings.md]:
```suggestion
* the new `defaults` object, which contains all the settings that should apply to
```

File: doc/user-docs/UsingJsonSettings.md [hunk 3]
```diff
@@ -199,6 +199,119 @@ like to hide all the WSL profiles, you could add the following setting:
...
+  every profile.
+
+What if I wanted a profile to have a different value for a property other than
+the default? Simply set the property in the profile's entry to override the
+value from `defaults`. Let's say you want the `cmd` profile to have `Consolas`
```
Comment (DHowett-MSFT) [inline doc/user-docs/UsingJsonSettings.md]: Inconsistent formatting for font names. Shall we split the difference and always italicize them?

File: src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp [hunk 4]
```diff
@@ -7,6 +7,7 @@
 #include "../TerminalApp/CascadiaSettings.h"
 #include "JsonTestClass.h"
 #include <defaults.h>
+#include "../ut_app/TestDynamicProfileGenerator.h"
```
Comment (DHowett-MSFT) [inline src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp]: Should we converge these tests at some point?

Comment (zadjii-msft) [reply src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp]: There's #3838 already, which includes converging these two projects.

File: src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp [hunk 5]
```diff
@@ -1781,4 +1783,306 @@ namespace TerminalAppLocalTests
         VERIFY_ARE_EQUAL(5555, settings._profiles.at(3)._historySize);
     }
 
+    void SettingsTests::TestTerminalArgsForBinding()
```
Comment (DHowett-MSFT) [inline src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp]: what happened here? this seems unrelated to this PR.

File: src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp [hunk 6]
```diff
@@ -63,6 +64,10 @@ namespace TerminalAppLocalTests
         TEST_METHOD(TestCloseOnExitParsing);
         TEST_METHOD(TestCloseOnExitCompatibilityShim);
 
+        TEST_METHOD(TestLayerUserDefaultsBeforeProfiles);
+        TEST_METHOD(TestDontLayerGuidFromUserDefaults);
+        TEST_METHOD(TestLayerUserDefaultsOnDynamics);
+
         TEST_METHOD(TestTerminalArgsForBinding);
```
Comment (DHowett-MSFT) [inline src/cascadia/LocalTests_TerminalApp/SettingsTests.cpp]: oh i see, it came BACK

File: src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp [hunk 7]
```diff
@@ -524,6 +543,46 @@ Profile* CascadiaSettings::_FindMatchingProfile(const Json::Value& profileJson)
...
+        }
+    }
+
+    if (defaultSettings)
+    {
+        _userDefaultProfileSettings = defaultSettings;
+
+        // Remove the `guid` member from the default settings. That'll
+        // hyper-explode, so just don't let them do that.
+        _userDefaultProfileSettings.removeMember({ "guid" });
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp]: Shouldn't we also remove `hidden`?

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp]: I mean, people could totally just hide _all_ their profiles though, right? Then only unhide profiles if they want them visible (i.e hide _all_ WSL profiles, and only re-enable the cmd profile)

Comment (DHowett-MSFT) [reply src/cascadia/TerminalApp/CascadiaSettingsSerialization.cpp]: yeah i honestly think that's fine
